### PR TITLE
Device-side printf()

### PR DESCRIPTION
**Pull request by franz**
implements device-side printf.

I added an external project (MIT license) for %e / %f / %g conversions, but by default use libc via snprintf.

### CODE REVIEW

File: lib/kernel/tinyprintf.c [hunk 1]
```diff
@@ -1,8 +1,9 @@
...
-File: tinyprintf.c
+/* OpenCL built-in library: tiny_printf
 
 Copyright (C) 2004  Kustaa Nyholm
 
+Copyright (C) 2018  Michal Babej
+
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
```
Comment (pjaaskel) [inline lib/kernel/tinyprintf.c]: The LGPL might be a problem here. I think an interpretation is that it's OK only to link to a LGPL lib without infecting other codes with the license. Can we find a more liberal replacement?

File: lib/llvmopencl/Workgroup.cc [hunk 2]
```diff
@@ -236,6 +258,12 @@ Workgroup::runOnModule(Module &M)
       }
   }
 
+    // remove functions which were replaced by their printf-augmented variants
```
Comment (pjaaskel) [inline lib/llvmopencl/Workgroup.cc]: Please do not commit commented out code.

File: lib/kernel/tinyprintf.c (hunk 1 above)
Comment (franz) [reply lib/kernel/tinyprintf.c]: Actually, it's dual-licensed. http://www.sparetimelabs.com/tinyprintf/tinyprintf.php

> Note, as of 8.5.2012 the license has been changed to BSD so you are free to use the LGPL licensed from this page or the BSD licensed from the link above.

Comment (pjaaskel) [reply lib/kernel/tinyprintf.c]: Ah, right. I think you already mentioned that. Then please attach the BSD license to the file and remove the LGPL.

**Pull request merged**